### PR TITLE
[TTAHUB-2607] iPD course name administration page

### PR DESCRIPTION
**Pull request by nvms**
## Description of change

Adds additional UI for course CRUD.

## How to test

* Add a course.
* Filter the course list.
* Select a course to rename or delete it.

## Issue(s)

* https://jira.acf.gov/browse/TTAHUB-2607

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Before merge to main

- [ ] OHS demo complete
- [ ] Ready to create production PR

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/routes/courses/handlers.ts [hunk 1]
```diff
@@ -28,6 +32,84 @@ export async function allCourses(req: Request, res: Response) {
...
+      res.status(403).send('Forbidden');
+      return;
+    }
+
+    const course = await getById(Number(id));
+    if (!course) {
+      res.status(404).send('Course not found');
+      return;
+    }
+    await course.destroy();
```
Comment (GarrettEHill) [inline src/routes/courses/handlers.ts:106]: this will fail if the course is already used, and there for linked to a report, right?

File: frontend/src/pages/Admin/Courses.js [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
   const primaryIdColumn = 'course name';
   const typeName = 'courses';
   const apiPathName = 'courses';
 
-  useEffect(() => {
-    async function get() {
-      const response = await getCourses();
-      setCourses(response);
-    }
+  const refresh = useCallback(async () => {
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/Admin/Courses.js:15]: Should we be using useEffect here? Why the change?

**Review by AdamAdHocTeam — COMMENTED** [review]
Not sure how hard it would be to set the two bottom under the two above.
image

**Comment by AdamAdHocTeam** [comment]
> Not sure how hard it would be to set the two bottom under the two above (on smaller screens). image

NMV I had search context (duh)

File: frontend/src/pages/Admin/Courses.js (hunk 2 above)
Comment (nvms) [reply frontend/src/pages/Admin/Courses.js:15]: `refresh` is a function that is provided to a child component, so it should be memoized with `useCallback`.

File: frontend/src/pages/Admin/__tests__/CourseAdd.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+describe('CourseAdd', () => {
+  const mockRefresh = jest.fn();
+
+  beforeEach(() => {
+    fetchMock.reset();
+    render(<CourseAdd refresh={mockRefresh} />);
+  });
+
+  it('renders the CourseAdd component', () => {
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/Admin/__tests__/CourseAdd.js:17]: For coverage do you have a check for the ' Add course' button click?

File: frontend/src/pages/Admin/__tests__/CourseList.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  screen,
+  fireEvent,
+} from '@testing-library/react';
+import CourseList from '../CourseList';
+
+describe('CourseList', () => {
+  const courses = [
+    { id: 1, name: 'Introduction to Dogs' },
+    { id: 2, name: 'Advanced Penguins' },
+    { id: 3, name: 'Data Structures and Kangaroos' },
```
Comment (AdamAdHocTeam) [inline frontend/src/pages/Admin/__tests__/CourseList.js:13]: No how to make pancakes... lol

File: frontend/src/pages/Admin/CourseAdd.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  const create = async () => {
+    await createCourseByName(courseName);
+    refresh();
+    setCourseName('');
+  };
+
+  return (
+    <div>
+      <h2>Add a new course</h2>
```
[the remark this answers is not in the document]
Comment (nvms) [reply frontend/src/pages/Admin/CourseAdd.js:29]: Sounds good! 841e49359

File: frontend/src/pages/Admin/__tests__/CourseAdd.js (hunk 3 above)
Comment (nvms) [reply frontend/src/pages/Admin/__tests__/CourseAdd.js:21]: Just added this: 841e49359

File: src/routes/courses/handlers.ts [hunk 6]
```diff
@@ -28,6 +32,84 @@ export async function allCourses(req: Request, res: Response) {
   }
 }
 
+export async function getCourseById(req: Request, res: Response) {
```
[the remark this answers is not in the document]
Comment (nvms) [reply src/routes/courses/handlers.ts:35]: We do now! c365ee4759ca5e927b9b2b030d5ca4544cc86613

File: src/routes/courses/handlers.ts (hunk 1 above)
Comment (nvms) [reply src/routes/courses/handlers.ts]: Yeah, good point. Catching and responding appropriately now: 2539aa5e83ca7ed4207a11c34127c7af954ad8b9

File: src/routes/courses/handlers.ts [hunk 7]
```diff
@@ -28,6 +32,89 @@ export async function allCourses(req: Request, res: Response) {
...
+      res.status(403).send('Forbidden');
+      return;
+    }
+
+    const course = await getById(Number(id));
+    if (!course) {
+      res.status(404).send('Course not found');
+      return;
+    }
+    const updated = await course.update(req.body);
```
Comment (thewatermethod) [inline src/routes/courses/handlers.ts:62]: I see there is a mapsTo column on the Courses table. I guess we aren't using it/following that pattern? 
Not a leading question, just curious.

Comment (nvms) [reply src/routes/courses/handlers.ts]: Thanks for pointing this out. Using it now: 9f4ebfbd6

File: frontend/src/pages/Admin/CourseList.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        id="courses-filter"
+        name="courses-filter"
+        key="courses-filter"
+        value={filter}
+        onChange={onFilterChange}
+        className="margin-bottom-2"
+      />
+      {filteredCourses.map((course) => (
+        <div key={course.id} className="margin-bottom-1">
+          <a href={`/admin/course/${course.id}`}>{course.name}</a>
```
Comment (thewatermethod) [inline frontend/src/pages/Admin/CourseList.js:36]: These should be <Link>'s so they use the react router navigation

Comment (nvms) [reply frontend/src/pages/Admin/CourseList.js]: 👍 d50717742

**Review by thewatermethod — APPROVED** [review]
